[RFC] API change: Device - Const-ify device driver instances.

tbursztyka opened the pull request:
This is a fix for #27399 

(Part of #22941, take a look if unaware of the device driver model rework)

Making all device instances constant.

Note: ALL checkpatch warnings are due to original code and not from this PR, so nevermind.

Commit:
device: Revise how initialization status is being handled

In order to make all device instances constant, driver_api pointer is
not set to NULL anymore if initialization failed.

Instead, have a bitfield dedicated to it.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
device: Const-ify all device driver instance pointers

Now that device_api attribute is unmodified at runtime, as well as all
the other attributes, it is possible to switch all device driver
instance to be constant.

A coccinelle rule is used for this:

@r_const_dev_1
  disable optional_qualifier
@
@@
-struct device *
+const struct device *

@r_const_dev_2
 disable optional_qualifier
@
@@
-struct device * const
+const struct device *

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: gpio: Use DEVICE_AND_API_INIT for mmio32 driver

Let's set the api at built time, or this will create a bug once device
instance pointers become constant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
device: Instantiate all device instance as constant

As they are set at build time.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pabigot commented:
Existing code that passes a device pointer as an argument to a function that receives a `void *` parameter is going to be a huge problem.  The `const` qualifier must be cast away in the ISR infrastructure, as well as many thread entry functions and other callback user-data parameters.  At this time the necessary conversions don't appear to be automatable.

pabigot committed:
userspace: update k_object API to support immutable objects

The k_object API associates mutable state structures with known kernel
objects to support userspace.  The kernel objects themselves are not
modified by the API, and in some cases (e.g. device structures) may be
const-qualified.  Update the API so that pointers to these const
kernel objects can be passed without casting away the const qualifier.

Fixes #27399

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

Commit:
isr: Normalize usage of device instance through ISR

The goal of this patch is to replace the 'void *' parameter by 'struct
device *' if they use such variable or just 'const void *' on all
relevant ISRs

This will avoid not-so-nice const qualifier tweaks when device instances
will be constant.

Note that only the ISR passed to IRQ_CONNECT are of interest here.

In order to do so, the script fix_isr.py below is necessary:

from pathlib import Path
import subprocess
import pickle
import mmap
import sys
import re
import os

cocci_template = """
@r_fix_isr_0
@
type ret_type;
identifier P;
identifier D;
@@
-ret_type <!fn!>(void *P)
+ret_type <!fn!>(const struct device *P)
{
 ...
(
 const struct device *D = (const struct device *)P;
|
 const struct device *D = P;
)
 ...
}

@r_fix_isr_1
@
type ret_type;
identifier P;
identifier D;
@@
-ret_type <!fn!>(void *P)
+ret_type <!fn!>(const struct device *P)
{
 ...
 const struct device *D;
 ...
(
 D = (const struct device *)P;
|
 D = P;
)
 ...
}

@r_fix_isr_2
@
type ret_type;
identifier A;
@@
-ret_type <!fn!>(void *A)
+ret_type <!fn!>(const void *A)
{
 ...
}

@r_fix_isr_3
@
const struct device *D;
@@
-<!fn!>((void *)D);
+<!fn!>(D);

@r_fix_isr_4
@
type ret_type;
identifier D;
identifier P;
@@
-ret_type <!fn!>(const struct device *P)
+ret_type <!fn!>(const struct device *D)
{
 ...
(
-const struct device *D = (const struct device *)P;
|
-const struct device *D = P;
)
 ...
}

@r_fix_isr_5
@
type ret_type;
identifier D;
identifier P;
@@
-ret_type <!fn!>(const struct device *P)
+ret_type <!fn!>(const struct device *D)
{
 ...
-const struct device *D;
...
(
-D = (const struct device *)P;
|
-D = P;
)
 ...
}
"""

def find_isr(fn):
    db = []
    data = None
    start = 0

    try:
        with open(fn, 'r+') as f:
            data = str(mmap.mmap(f.fileno(), 0).read())
    except Exception as e:
        return db

    while True:
        isr = ""
        irq = data.find('IRQ_CONNECT', start)
        while irq > -1:
            p = 1
            arg = 1
            p_o = data.find('(', irq)
            if p_o < 0:
                irq = -1
                break;

            pos = p_o + 1

            while p > 0:
                if data[pos] == ')':
                    p -= 1
                elif data[pos] == '(':
                    p += 1
                elif data[pos] == ',' and p == 1:
                    arg += 1

                if arg == 3:
                    isr += data[pos]

                pos += 1

            isr = isr.strip(',\\n\\t ')
            if isr not in db and len(isr) > 0:
                db.append(isr)

            start = pos
            break

        if irq < 0:
            break

    return db

def patch_isr(fn, isr_list):
    if len(isr_list) <= 0:
        return

    for isr in isr_list:
        tmplt = cocci_template.replace('<!fn!>', isr)
        with open('/tmp/isr_fix.cocci', 'w') as f:
            f.write(tmplt)

        cmd = ['spatch', '--sp-file', '/tmp/isr_fix.cocci', '--in-place', fn]

        subprocess.run(cmd)

def process_files(path):
    if path.is_file() and path.suffix in ['.h', '.c']:
        p = str(path.parent) + '/' + path.name
        isr_list = find_isr(p)
        patch_isr(p, isr_list)
    elif path.is_dir():
        for p in path.iterdir():
            process_files(p)

if len(sys.argv) < 2:
    print("You need to provide a dir/file path")
    sys.exit(1)

process_files(Path(sys.argv[1]))

And is run: ./fix_isr.py <zephyr root directory>

Finally, some files needed manual fixes such.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: ethernet: Remove ugly usage of a goto in enc*j* drivers

This is a working code, but it's harder to read. And for some reason
makes some semantic patches of coccinelle running forever.

So refactoring it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
scripts: coccinelle: Scripts for finding wrong device instance usage

2 scripts are provided.
- find_functions.cocci (name probably sucks...)
- find_dev_usage.cocci (ditto...)

find_functions.cocci can patch files where device instance are not
const.

Then it is used to generate the function database:

./scripts/coccicheck --mode=report --jobs=1 \
--cocci=scripts/coccinelle/find_functions.cocci \
--sp-flag="--include-headers" ./

Then, find_dev_usage.cocci will check if the const qualifier is, or
might be lost in a function call.

For instance:
./scripts/coccicheck --mode=report --jobs=1 \
--cocci=scripts/coccinelle/find_dev_usage.cocci \
--sp-flag="--include-headers" drivers/i2c

Which will output a WARNING on non-zephyr functions calls with a device:

./drivers/i2c/i2c_cc13xx_cc26xx.c:393:5-8: WARNING: in i2c_cc13xx_cc26xx_pm_control calling cb param with dev, check if const qualifier is not lost
./drivers/i2c/i2c_mcux_lpi2c.c:205:40-43: WARNING: in mcux_lpi2c_init calling LPI2C_MasterTransferCreateHandle param with dev, check if const qualifier is not lost
./drivers/i2c/i2c_nrfx_twi.c:258:5-8: WARNING: in twi_nrfx_pm_control calling cb param with dev, check if const qualifier is not lost
./drivers/i2c/i2c_nrfx_twi.c:202:22-25: WARNING: in init_twi calling nrfx_twi_init param with dev, check if const qualifier is not lost
./drivers/i2c/i2c_mcux.c:187:38-41: WARNING: in i2c_mcux_init calling I2C_MasterTransferCreateHandle param with dev, check if const qualifier is not lost
./drivers/i2c/i2c_mcux_flexcomm.c:184:43-46: WARNING: in mcux_flexcomm_init calling I2C_MasterTransferCreateHandle param with dev, check if const qualifier is not lost
./drivers/i2c/i2c_nrfx_twim.c:232:5-8: WARNING: in twim_nrfx_pm_control calling cb param with dev, check if const qualifier is not lost
./drivers/i2c/i2c_nrfx_twim.c:174:8-11: WARNING: in init_twim calling nrfx_twim_init param with dev, check if const qualifier is not lost
./drivers/i2c/i2c_rv32m1_lpi2c.c:246:6-9: WARNING: in rv32m1_lpi2c_init calling LPI2C_MasterTransferCreateHandle param with dev, check if const qualifier is not lost

Or:
./scripts/coccicheck --mode=report --jobs=1 \
--cocci=scripts/coccinelle/find_dev_usage.cocci \
--sp-flag="--include-headers" drivers/ieee802154

Which will output an ERROR on using a zephyr function that looses the
const qualifier:

drivers/ieee802154/ieee802154_rf2xx.c:778:3-6: ERROR: in rf2xx_init calling k_thread_create param with dev, loosing const qualifier, please wrap
drivers/ieee802154/ieee802154_nrf5.c:477:19-22: ERROR: in nrf5_init calling k_thread_create param with dev, loosing const qualifier, please wrap
drivers/ieee802154/ieee802154_cc1200.c:819:3-6: ERROR: in cc1200_init calling k_thread_create param with dev, loosing const qualifier, please wrap
drivers/ieee802154/ieee802154_mcr20a.c:1443:3-6: ERROR: in mcr20a_init calling k_thread_create param with dev, loosing const qualifier, please wrap
drivers/ieee802154/ieee802154_cc2520.c:1116:3-6: ERROR: in cc2520_init calling k_thread_create param with dev, loosing const qualifier, please wrap
drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:439:32-35: ERROR: in ieee802154_cc13xx_cc26xx_data_init calling k_thread_create param with dev, loosing const qualifier, please wrap

ISSUE:
- Is it possible to run a set of rules first on all the code, and then
  another set, both sets being in the same .cocci file?
  Would be nice to have all at once.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: i2c: Fix device instance const qualifier loss

In all of these drivers, passing the device's data was sufficient as
only the data is being used by the HAL callback function then.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: can: Fix device instance const qualifier loss in MCUX Flexcan

It is necessary to wrap the device pointer into data, and pass the data
to the HAL callback function.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: dma: Fix device instance const qualifier loss in NIOS2 driver

Passing the device's config was necessary, and wrapping device inside
it.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: gpio: Fix device instance const qualifier loss

It is necessary to wrap the device pointer into data.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: watchdog: Remove useless Kconfig entry

The function it would build is nowhere to be seen.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: spi: Fix device instance const qualifier loss

Some needed to wrap the device pointer into device's data, where others
needed only device's data to be passed to HAL callback function.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: flash: Fix device instance const qualifier loss in NRF QSPI

Passing the device's data is sufficient to be used by the HAL callback
function.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: video: Fix device instance const qualifier loss in MCUX CSI

It is necessary to wrap the device pointer into data.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: ieee802154: Fix device instance const qualifier loss

In all of these drivers, passing the device's data was sufficient as
only the data is being used by thread.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: ethernet: Fix device instance const qualifier loss

Passing driver's data to k_thread is sufficient for mcux.
On enc424j600, however, the device pointer is needed and thus is wrapped
into its data. But there seems to be a possible optimisation: all local
spi related function in fact only needs the device's data (context) and
so changing all spi related function to take the context would remove
the need to wrap the device pointer into its data.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: adc: Fix device instance const qualifier loss

It is necessary to wrap the device pointer into data.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: wifi: Fix device instance const qualifier loss in ESP

Passing the device's data was sufficient as only the data is being used
by the thread.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: console: Fix device instance const qualifier loss in IPM driver

Passing the device's data and config is sufficient for the thread.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: led: Fix device instance const qualifier loss in HT16K33

It is necessary to wrap the device pointer into data.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: kscan: Fix device instance const qualifier loss

SDL one needs to wrap device pointer into its data and xec do not need
to pass any argument to its thread.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: serial: Fix device instance const qualifier loss

It is necessary to wrap the device pointer into data.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: sensor: Fix device instance const qualifier loss

It is necessary to wrap the device pointer into data.
Which was done already on most of them when global trigger is enabled.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
subsys: logging: Fix device instance const qualifier loss

Uart device is unique and thus does not need to be passed through the
logging context. Only the assert() requires a void * cast.

On log_output side, the device usage was removed. It is actually unclear
why it has been set like this since depending on the context, it can be
anything and not specifically a device.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
subsys: net: Fix device instance const qualifier loss in socket tls

Entropy device, as being unique here, does not need to be passed through
mbedtls and can instead set static locally.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
subsys: canopen: Add a wrapper to avoid passing device to CO_init

Add a dedicated structure instead.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples: Fix device instance const qualifier loss

Keep the device pointer locally.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
irq: Change offload API to take a constant parameter

All ISRs are meant to take a const struct device pointer, but to
simplify the change let's just move the parameter to constant and that
should be fine.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch: Apply IRQ offload API change

Switching to constant parameter.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
kernel: Apply IRQ offload API change

Switching to constant parameter.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests: Apply IRQ offload API change

Switching to constant parameter.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
irq: Change dynamic API to take a constant parameter

All ISRs are meant to take a const struct device pointer, but to
simplify the change let's just move the parameter to constant and that
should be fine.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch: Apply dynamic IRQ API change

Switching to constant parameter.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests: Apply dynamic IRQ API change

Switching to constant parameter.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards: Apply IRQ offload API change

Switching to constant parameter on ISR handler.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests: uart_asyn: Simplify how the UART device is provided

Just get it once as device_get_binding() is a bit greedy.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: Manual const-ification of device driver instance

These are all the case that coccinelle cannot find as they are inside
macro declarations.

Fixed via:

git grep -rlz -E "\(struct device \*" |
	xargs -0 sed -i 's/(struct device/(const struct device/g'

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
Modules need to be updated now: tee/debug/segger and crypto/mbedtls
That's why CI is failing at the moment.

nzmichaelh commented:
Hi @tbursztyka sorry if this has already been discussed, but is it OK to split this into multiple smaller PRs?  This is 10k lines of change which is hard to review and hard to merge without conflicts.

For example, could all of the UART related changes be split into a PR?

tbursztyka commented:
@nzmichaelh see #26426 (there are 3 PRs it's based on, listed on topic message of this PR. I'll add the title of each PR)

pabigot commented on arch/arc/core/arc_smp.c:
```diff
@@ -108,7 +108,7 @@ void arch_sched_ipi(void)
 	}
 }
 
-static int arc_smp_init(struct device *dev)
```
Comment is on the commit titled:

    device: Const-ify all device driver instance pointers

which produced this change.  The provided Coccinelle script:

```
@r_const_dev_basic
  disable optional_qualifier
@
@@
-struct device *
+const struct device
```
is wrong in two ways:
* it changes a pointer to `struct device` to a `const struct device` object
* it fails to recognize a certain style that puts a `const` qualifier on the variable declaration.

For completeness it should be something like this:
```
@
disable optional_qualifier
@
@@
+const
 struct device *

@
disable optional_qualifier
@
@@
+const
 struct device * const
```

tbursztyka replied:
sure

pabigot commented on drivers/kscan/kscan_mchp_xec.c:
```diff
@@ -51,7 +51,7 @@ struct kscan_xec_data {
 	uint8_t scan_cycles_idx;
 	kscan_callback_t callback;
-	struct k_thread thread;
+	struct k_thread threa;
```
This looks like a typo.

tbursztyka replied:
good point

pabigot reviewed (dismissed):
This review excludes the commits related to changing callbacks, which were reviewed separately in their base PRs.

On the whole I think this is good.  But it's also a huge set of changes, and I'm not sure we'll find everything that got broken.  That's ok, we just need to recognize that these problems may occur.

Text changes requested for commits:

* device: Instan~c~*t*iate all device instance*s* as constant

The commit message of

    scripts: coccinelle: Scripts for finding wrong device instance usage

suggests it's still a draft.  I have not yet looked at the details of these as they seem to be used only for the transformation in this PR; not clear if they're intended to be merged.

I think changing the irq_offload and other API to take a `const void *` is a nice solution, though am a bit concerned that casts might be removing the `const` qualifier which risks undefined behavior.  I have not closely examined all the cases to see whether that happens (another thing that could be done with Coccinelle).

tbursztyka commented on include/linker/common-ram.ld:
```diff
@@ -14,18 +14,27 @@
+#define DEVICE_BITFIELD_SIZE	(((DEVICE_COUNT + 31) / 32) * 4)
+
+#define DEVICE_INIT_STATUS_BITFIELD() 		\
+		FILL(0xFF);			\
```
I followed the legacy behavior: driver_api is set at build time, and unset if init fails. Could be other way round (though it means it will set the bit for each and every device, where here it unsets if only init fails. probably not a performance hit though)

tbursztyka replied:
I'll switch to the other way round, which makes more sense.

tbursztyka commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -167,13 +167,16 @@ static int gpio_sam0_pin_interrupt_configure(const struct device *dev,
 					     enum gpio_int_trig trig)
 {
 	const struct gpio_sam0_config *config = DEV_CFG(dev);
+	struct gpio_sam0_data *const data = DEV_DATA(dev);
```
yes I had to go through a grep and manually change these afterwards. Not nice at all indeed.

Commit:
west: Update module's revision

These revisions have the right fix for device const-ification.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: Apply dynamic IRQ API change

Switching to constant parameter.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
doc: Add a device qualifier change entry to release-notes for 2.4 version

Documenting the change.

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
Ok, PR seems ready.
The are 2 build failures unrelated to that PR, for which I opened issues #27379 and #27463 
Other failures come (time to time) from qemu_arc_hs/qemu_arc_em on tests that never fails on my side. I guess it's due to timing issues on CI so not relevant here.

@carlescufi , @pabigot , @nashif , @galak  (among others) please (re)review.

pabigot commented:
It looks like the content in patch 5 ("device: Const-ify all device driver instance pointers") has been modified beyond what Coccinelle does.  The rule removes `const` qualifiers were from variable definitions, unlike the approach I proposed, which isn't really an issue though it's an unrelated undocumented change.

But ReStructuredText documentation changes which couldn't have been done by coccinelle are also in it, as well as changes to macros that Coccinelle doesn't see.  Also there are instances of `struct device const` where Coccinelle produces `const struct device`.

Patch 4 description is inaccurate: only two device init functions are changed in it.  It appears to be necessary because the Coccinelle script in patch 5 was not re-invoked on an updated master branch that included those files: if it had that patch would not be necessary.

It would be nicer to have converted callbacks to take `struct device *` instead of `void *` first, where necessary, separate from the switch to `const struct device`.  Then more material could have been handled by Coccinelle.

So the changes I requested were made and my NAK removed, but I'm less comfortable with this patch set than last review because it's not clear what was done automatically and what was done manually. 10K edits to ~1500 files isn't something that can be reviewed when things are mixed this way.

carlescufi commented:
API meeting 11th August 2020:

The issues that were raised during the meeting can be summarized as:

- Whenever an automated tool is run in order to make changes to the codebase, that set of changes should be the only content in a commit. This will break bisectability, but will also ensure that the changes can be reproduced by a reviewer (as @pabigot indeed did, and spotted this very issue) by running the exact same tool with the same parameters. `git bisect skip` can be used to work around.

tbursztyka commented:
@pabigot Probably got lost rebase after rebase where I pusch changes in the wrong commit. I'll try to see if I can fix this. (and rebase again against master)

pabigot commented:
There are still commits in here that cannot be reproduced by invoking the commands identified in the commit message.

I would like to see a plan for coordinating when and how this gets merged relative to other changes in master.  I'd rather not see more failures like #27594.

The only plan I can see being feasible is locking down master and doing a rebase with a pause before each commit that includes automated changes.  At that point run the corresponding automated process and commit the results, then continue the rebase skipping the old commit.  As far as I know that can't be done if automated and manual changes appear in the same commit, because the manual changes will be lost.

While I was reviewing a patch got merged that breaks a rebase in exactly this way.

There are also a few commits that are perhaps worth of being merged, but IMO don't belong as part of this PR as they're not related to the primary goal nor do they appear to interfere with reaching that goal.

Notes on specific commits below.

> [PATCH 05/42] device: Const-ify all device driver instance pointers

still contains changes that are not produced by Coccinelle, though maybe there are fewer. Some are documentation, some appear to be formatting, some are things that Coccinelle couldn't find.

> [PATCH 12/42] drivers: watchdog: Remove useless Kconfig entry

This is clearly a mistake from #11442 where the call should have been updated to `wdog_cmsdk_apb_setup()`.  The patch should be pulled out to a separate PR: it's not related to the purpose of this PR.

> [PATCH 38/42] isr: Normalize usage of device instance through ISR

Contains changes not produced by the script provided in the commit description.

> PATCH 39/42] tests: uart_asyn: Simplify how the UART device is provided

Doesn't seem related to purpose of PR.

> [PATCH 40/42] drivers: Manual const-ification of device driver instance

Whitespace changes not generated by the documented command.

andrewboie commented:
> > [PATCH 05/42] device: Const-ify all device driver instance pointers
> 
> still contains changes that are not produced by Coccinelle, though maybe there are fewer. Some are documentation, some appear to be formatting, some are things that Coccinelle couldn't find.

I would like to step back a second and ask is this **really** a requirement?

From my point of view, spending tons of time to get coccinelle scripts to catch every last corner case is letting the perfect be the enemy of the good, and presumes that these scripts get re-used far more than the actually do in practice.

pabigot commented:
> > still contains changes that are not produced by Coccinelle, though maybe there are fewer. Some are documentation, some appear to be formatting, some are things that Coccinelle couldn't find.
> 
> I would like to step back a second and ask is this **really** a requirement?
> 
> From my point of view, spending tons of time to get coccinelle scripts to catch every last corner case is letting the perfect be the enemy of the good, and presumes that these scripts get re-used far more than the actually do in practice.

@andrewboie I agree entirely.  I wouldn't even bother adding the scripts to discover new instances.  I can't see them being used much.

But rebasing this PR is a nightmare as long as the changes they produce are merged with manual changes in the same commit.  You can do it, but you have to hand-edit all the places where automated changes conflict with changes that were merged since the previous base commit.  That hand editing is error prone.  Give it a try.

On the other hand it took 20 minutes to split the 3 commits that have automated changes into two: one that has the exact output of the automation, the other that has the remaining changes.  Half that time was dealing with the conflicts necessary to fix the backslash alignment in long macros, which was manual but straightforward (edit each file, taken the second version of each conflict).  Verify that [the result](https://github.com/pabigot/zephyr/commits/xpr/24873a) matches the original.

Once done, it took four minutes to rebase this on current master by stopping at each commit that's purely automation, discarding it, re-running the automation, and committing the result before continuing.  Most of *that* time was waiting for `spatch` to build the patch file for the first automated change.  Compare [that](https://github.com/pabigot/zephyr/commits/xpr/24873b) against the previous one and verify that the diff is the same as the diff between the master versions the two were based on.

The next rebase should also take about four minutes.  And the one after that.  And so on, for however many rebases are necessary until we can finally get this merged.

At any rate, all I can do is suggest what seems a better way to do this.  It's not my place to insist on it being done that way, and in fact I never did suggest it was a requirement.

andrewboie commented:
> It's not my place to insist on it being done that way, and in fact I never did suggest it was a requirement.

It's not always perfectly clear in your multi-point reviews which criteria are a gating factor. Glad we cleared that up.

> The next rebase should also take about four minutes. And the one after that. And so on, for however many rebases are necessary until we can finally get this merged.

There are human factors that could be leveraged for this.

Instead of rebasing over and over again, let this air out as-is, and then once there is sufficient review consensus, have the people who are empowered to merge code hold off on merging other conflicting PRs for a day or two until a final rebase and validation pass can happen. In the unlikely event something slips through, fix it in a follow-up. This has worked for us in the past.

tbursztyka commented:
Opened a new issue on one bug unrelated to this pr: #27634

tbursztyka commented:
>     [PATCH 05/42] device: Const-ify all device driver instance pointers
>
> still contains changes that are not produced by Coccinelle, though maybe there are fewer. Some are documentation, some appear to be formatting, some are things that Coccinelle couldn't find.

I'll check for the obvious ones, but I won't be exhaustive. (Actually I don't see this as a big issue).

>    [PATCH 12/42] drivers: watchdog: Remove useless Kconfig entry
>
>This is clearly a mistake from #11442 where the call should have been updated to wdog_cmsdk_apb_setup(). The patch should be pulled out to a separate PR: it's not related to the purpose of this PR.

CI has been finding many unrelated issues like this one on this PR. I can pull it out, but CI will again fail on that here. It will again delay over and over this PR...  

>   [PATCH 38/42] isr: Normalize usage of device instance through ISR
>
>Contains changes not produced by the script provided in the commit description.

The commit messages does tell about manual fixes in addition of the cocci stuff.

>    PATCH 39/42] tests: uart_asyn: Simplify how the UART device is provided
>
> Doesn't seem related to purpose of PR.

sure but, let's just keep it there. It's no harm. I'll remove the "Fixes:..." from the commit message though.

>    [PATCH 40/42] drivers: Manual const-ification of device driver instance
>
> Whitespace changes not generated by the documented command.

I guess I did these to avoid the +80chars limit issue.

Commit:
doc: Apply constant qualifier on device instance where relevant

s/struct device */const struct device *

Fixes #27399

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -440,12 +438,13 @@ static int qspi_nrfx_configure(const struct device *dev)
 		return -ENXIO;
 	}
 
+	struct qspi_nor_data *const driver_data = dev->driver_data;
```
This variable should be named `dev_data` instead of `driver_data` (that's what is used in the call below and also in the event handler).
Also this `const` here does not seem to bring any value, and it's not consistent with other declarations of such variables, so I'd suggest to remove it.

carlescufi commented:
@nashif @anangl @henrikbrixandersen @MaureenHelm @mnkp @nordic-krch @galak @dcpleung a review would be appreciated in order to get this into 2.4

carlescufi reviewed (approved):
In general this looks good to me.

The only nit that I wanted to bring up is the placement of the `const` keyword: this PR uses `const struct device *`, but there is also the option of using `struct device const *`, which is equivalent and can be read backwards: "pointer to constant device structure". That said the numbers do favor the current state of the PR:

```
carles@zephyr-nrf ~/src/zephyr/zephyr (master *)
$ git grep "const struct" | wc -l
7896

carles@zephyr-nrf ~/src/zephyr/zephyr (master *)
$ git grep "const \*" | wc -l
174
```

And consistency comes before everything else, so approving this.

chrta commented:
The following fails to build for me:
```
$ sanitycheck -p native_posix -s tests/kernel/threads/thread_apis/kernel.threads.apis

...

/home/user/projects/zephyrproject/zephyr/tests/kernel/threads/thread_apis/src/test_threads_cancel_abort.c: In function ‘entry_abort_isr’:
/home/user/projects/zephyrproject/zephyr/tests/kernel/threads/thread_apis/src/test_threads_cancel_abort.c:211:14: error: passing argument 1 of ‘irq_offload’ from incompatible pointer type [-Werror=incompatible-pointer-types]
  irq_offload(offload_func, k_current_get());
              ^~~~~~~~~~~~
In file included from /home/user/projects/zephyrproject/zephyr/include/sys/arch_interface.h:34:0,
                 from /home/user/projects/zephyrproject/zephyr/include/arch/cpu.h:12,
                 from /home/user/projects/zephyrproject/zephyr/include/kernel_includes.h:38,
                 from /home/user/projects/zephyrproject/zephyr/include/kernel.h:17,
                 from /home/user/projects/zephyrproject/zephyr/include/zephyr.h:18,
                 from /home/user/projects/zephyrproject/zephyr/subsys/testsuite/ztest/include/ztest.h:58,
                 from /home/user/projects/zephyrproject/zephyr/tests/kernel/threads/thread_apis/src/test_threads_cancel_abort.c:7:
/home/user/projects/zephyrproject/zephyr/include/irq_offload.h:33:6: note: expected ‘irq_offload_routine_t {aka void (*)(const void *)}’ but argument is of type ‘void (*)(void *)’
 void irq_offload(irq_offload_routine_t routine, const void *parameter);
      ^~~~~~~~~~~
cc1: all warnings being treated as errors
```

SDK Version 0.11.1 on ubuntu 18.04

tbursztyka commented:
@chrta indeed, new additions that I did not see while rebasing

MaureenHelm commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -1017,7 +1015,7 @@ static void eth_mcux_init(const struct device *dev)
 	};
 #if defined(CONFIG_PTP_CLOCK_MCUX)
-	uint8_t ptp_multicast[6] = { 0x01, 0x80, 0xC2, 0x00, 0x00, 0x0E };
+	uint8_t ptp_multicast[6] = { 0x01, 0x80, 0xC2, 0x00, 0x00, 0x0E ;}
```
this doesn't look right

tbursztyka replied:
indeed

MaureenHelm commented on drivers/sensor/lis2ds12/lis2ds12.c:
```diff
@@ -184,13 +184,17 @@ static int lis2ds12_sample_fetch(const struct device *dev,
 		break;
 #if defined(CONFIG_LIS2DS12_ENABLE_TEMP)
 	case SENSOR_CHAN_DIE_TEMP:
-		lis2ds12_sample_fetch_temp(dev);
+		/* ToDo:
+		lis2ds12_sample_fetch_temp(dev)
+		*/
 		break;
 #endif
 	case SENSOR_CHAN_ALL:
 		lis2ds12_sample_fetch_accel(dev);
 #if defined(CONFIG_LIS2DS12_ENABLE_TEMP)
-		lis2ds12_sample_fetch_temp(dev);
+		/* ToDo:
+		lis2ds12_sample_fetch_temp(dev)
+		*/
```
todo?

tbursztyka replied:
lis2ds12_sample_fetch_temp() does not exist anywhere, and was giving me build error. That could be an externalized patch though

MaureenHelm replied:
@avisconti please review

MaureenHelm commented on west.yml:
```diff
@@ -56,7 +56,7 @@ manifest:
       path: modules/hal/infineon
     - name: hal_nordic
-      revision: 4be2a1eadc99511e0271b6de6e0be459a4cf0cef
+      revision: pull/48/head
```
please don't forget to update these to new SHAs

tbursztyka replied:
I'll update it once these pull requests get merged yes, so a tiny bit before this PR gets merged then.

MaureenHelm reviewed (changes requested):
looks good, just a couple nits. thanks for driving this @tbursztyka

henrikbrixandersen reviewed (approved):
I have reviewed a good portion of this patch with main focus on the files, I have touched in the past. Looks good to me.

vanti reviewed (approved):
Verified on SimpleLink.

carlescufi commented:
west.yml SHAs double-checked, merging the PR
